File: c++/linx/gpio.cpp
// General Purpose Input/Output Pin abstract interface module

// Copyright (C)2016, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <map>

#include <errno.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"
#include "gpio.h"

// Local state variables

typedef std::map<uint8_t, GPIO_Interface_Ptr> GPIO_Map_t;

static GPIO_Map_t PinTable;

//***************************************************************************

static void GetGPIOChannels(LINX_command_t *cmd, LINX_response_t *resp, int32_t *error)
{
  unsigned ChannelIndex = 0;

  PREPARE_RESPONSE;
  CHECK_COMMAND_SIZE(7, 7);

  for (GPIO_Map_t::iterator it=PinTable.begin(); it!=PinTable.end(); ++it)
    resp->Data[ChannelIndex++] = it->first;

  resp->PacketSize += ChannelIndex;
  resp->Status = L_OK;
  *error = 0;
}

//***************************************************************************

static void GPIOConfigure(LINX_command_t *cmd, LINX_response_t *resp, int32_t *error)
{
  unsigned MaxPins = PinTable.size();
  unsigned NumPins = cmd->Args[0];
  uint8_t *PinNums = &cmd->Args[1];
  unsigned i;

  PREPARE_RESPONSE;
  CHECK_COMMAND_SIZE(10, 9 + MaxPins + MaxPins / 8);
  CHECK_COMMAND_SIZE(9 + NumPins + NumPins / 8, 9 + NumPins + NumPins / 8);

  // Process the GPIO pin list

  for (i = 0; i < NumPins; i++)
  {
    int32_t p = PinNums[i];
    int32_t d = (cmd->Args[NumPins + 1 + i / 8] >> (i % 8)) & 0x01;
    GPIO_Interface_Ptr pin;

    // Lookup pin info

    try
    {
      pin = PinTable[p];
    }

    catch (int e)
    {
      resp->Status = L_UNKNOWN_ERROR;
      *error = ENODEV;
      return;
    }

    // Configure the GPIO pin

    pin->configure(d, error);
    if (*error)
    {
      resp->Status = L_UNKNOWN_ERROR;
      return;
    }

    pin->IsOutput = d;
  }

  resp->Status = L_OK;
  *error = 0;
}

//***************************************************************************

static void GPIOWrite(LINX_command_t *cmd, LINX_response_t *resp, int32_t *error)
{
  unsigned MaxPins = PinTable.size();
  unsigned NumPins = cmd->Args[0];
  uint8_t *PinNums = &cmd->Args[1];
  unsigned i;

  PREPARE_RESPONSE;
  CHECK_COMMAND_SIZE(10, 9 + MaxPins + MaxPins / 8);
  CHECK_COMMAND_SIZE(9 + NumPins + NumPins / 8, 9 + NumPins + NumPins / 8);

  // Process the GPIO pin list

  for (i = 0; i < NumPins; i++)
  {
    int32_t p = PinNums[i];
    int32_t v = (cmd->Args[NumPins + 1 + i / 8] >> (i % 8)) & 0x01;
    GPIO_Interface_Ptr pin;

    // Lookup pin info

    try
    {
      pin = PinTable[p];
    }

    catch (int e)
    {
      resp->Status = L_UNKNOWN_ERROR;
      *error = ENODEV;
      return;
    }

    // Configure the GPIO pin as an output, if necessary

    if (!pin->IsOutput)
    {
      pin->configure(true, error);
      if (*error)
      {
        resp->Status = L_UNKNOWN_ERROR;
        return;
      }

      pin->IsOutput = true;
    }

    // Write to the GPIO pin

    pin->write(v, error);
    if (*error)
    {
      resp->Status = L_UNKNOWN_ERROR;
      return;
    }
  }

  resp->Status = L_OK;
  *error = 0;
}

//***************************************************************************

static void GPIORead(LINX_command_t *cmd, LINX_response_t *resp, int32_t *error)
{
  unsigned MaxPins = PinTable.size();
  unsigned NumPins = cmd->PacketSize - 7;
  uint8_t *PinNums = &cmd->Args[0];
  unsigned i;

  PREPARE_RESPONSE;
  CHECK_COMMAND_SIZE(8, 7 + MaxPins);
  CHECK_COMMAND_SIZE(7 + NumPins, 7 + NumPins);

  // Process the GPIO pin list

  for (i = 0; i < NumPins; i++)
  {
    int32_t p = PinNums[i];
    int32_t v;
    GPIO_Interface_Ptr pin;

    // Lookup pin info

    try
    {
      pin = PinTable[p];
    }

    catch (int e)
    {
      resp->Status = L_UNKNOWN_ERROR;
      *error = ENODEV;
      return;
    }

    // Read the GPIO pin

    pin->read(&v, error);
    if (*error)
    {
      resp->Status = L_UNKNOWN_ERROR;
      return;
    }

    resp->Data[i / 8] |= (v << (7 - i % 8));
  }

  resp->PacketSize += NumPins/8 + 1;
  resp->Status = L_OK;
  *error = 0;
}

//***************************************************************************

// Add a GPIO pin to the pin table

void gpio_add_channel(uint8_t channel, GPIO_Interface_Ptr pin)
{
  PinTable[channel] = pin;
}

//***************************************************************************

// Register command handlers

void gpio_init(void)
{
  AddCommand(CMD_GET_GPIO_CHANNELS, GetGPIOChannels);
  AddCommand(CMD_GPIO_CONFIGURE, GPIOConfigure);
  AddCommand(CMD_GPIO_WRITE, GPIOWrite);
  AddCommand(CMD_GPIO_READ, GPIORead);
}

File: csharp/spiagent/gpio.cs
// Raspberry Pi LPC1114 I/O Processor Expansion Board SPI Agent Firmware
// GPIO services

// Copyright (C)2014-2018, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

using System;
using System.Collections.Generic;

namespace SPIAgent
{
    /// <summary>
    /// The GPIO (General Purpose Input Output) class implements LPC1114 digital input/output services.
    /// </summary>
    public class GPIO: IO.Interfaces.GPIO.Pin
    {
        /// <summary>
        /// LPC1114 GPIO pin data directions.
        /// </summary>
        public enum DIRECTION
        {
            INPUT,
            OUTPUT,
            SENTINEL
        }

        /// <summary>
        /// LPC1114 GPIO pin configuration modes.
        /// </summary>
        public enum MODE
        {
            INPUT,
            INPUT_PULLDOWN,
            INPUT_PULLUP,
            OUTPUT,
            OUTPUT_OPENDRAIN,
            SENTINEL
        }

        /// <summary>
        /// LPC1114 GPIO pin interrupt modes.
        /// </summary>
        public enum INTERRUPT
        {
            DISABLED,
            FALLING,
            RISING,
            BOTH,
            SENTINEL
        }

        private static HashSet<int> ValidPins = new HashSet<int>();
        private ITransport mytransport;
        private int mypin;
        private SPIAGENT_COMMAND_MSG_t cmd;
        private SPIAGENT_RESPONSE_MSG_t resp;

        /// <summary>
        /// Static constructor that populates the valid pins table.
        /// </summary>
        static GPIO()
        {
            ValidPins.Add(Pins.LPC1114_INT);
            ValidPins.Add(Pins.LPC1114_LED);
            ValidPins.Add(Pins.LPC1114_GPIO0);
            ValidPins.Add(Pins.LPC1114_GPIO1);
            ValidPins.Add(Pins.LPC1114_GPIO2);
            ValidPins.Add(Pins.LPC1114_GPIO3);
            ValidPins.Add(Pins.LPC1114_GPIO4);
            ValidPins.Add(Pins.LPC1114_GPIO5);
            ValidPins.Add(Pins.LPC1114_GPIO6);
            ValidPins.Add(Pins.LPC1114_GPIO7);
        }

        /// <summary>
        /// Constructor that configures a GPIO pin, given pin number, data direction, and initial state.
        /// </summary>
        /// <param name="spiagent">SPI Agent Firmware transport object.</param>
        /// <param name="pin">LPC1114 GPIO pin number.
        /// LPC1114_GPIO0 through LPC1114_GPIO7 are allowed.</param>
        /// <param name="dir">LPC1114 GPIO pin data direction.</param>
        /// <param name="state">LPC1114 GPIO pin initial state.
        /// For an output pin, sets output state to sourcing (1) or sinking (0).
        /// For an input pin, sets internal pull-up resistor (1) or pull-down resistor (0).</param>
        public GPIO(ITransport spiagent, int pin, DIRECTION dir, bool state)
        {
            // Validate parameters

            if (spiagent == null)
            {
                throw new NullReferenceException("SPI Agent Firmware transport handle is null");
            }

            if (!ValidPins.Contains(pin))
            {
                throw new ArgumentException("GPIO pin number is invalid");
            }

            if (dir >= DIRECTION.SENTINEL)
            {
                throw new ArgumentException("GPIO direction parameter is invalid");
            }

            cmd = new SPIAGENT_COMMAND_MSG_t();
            resp = new SPIAGENT_RESPONSE_MSG_t();

            // Build the command message

            cmd.pin = pin;

            switch (dir)
            {
                case DIRECTION.INPUT:
                    cmd.command = (int)Commands.SPIAGENT_CMD_CONFIGURE_GPIO_INPUT;
                    cmd.data = state ? 1 : 0;
                    break;

                case DIRECTION.OUTPUT:
                    cmd.command = (int)Commands.SPIAGENT_CMD_CONFIGURE_GPIO_OUTPUT;
                    cmd.data = state ? 1 : 0;
                    break;
            }

            // Dispatch the command

            spiagent.Command(cmd, ref resp);

            // Handle errors

            if (resp.error != 0)
            {
                throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
            }

            mytransport = spiagent;
            mypin = pin;
        }

        /// <summary>
        /// Constructor that configures a GPIO pin, given pin number and mode.
        /// Use this constructor if you need to configure a high-impedance input or an open-drain output.
        /// </summary>
        /// <param name="spiagent">SPI Agent Firmware transport object.</param>
        /// <param name="pin">LPC1114 GPIO pin number.
        /// LPC1114_GPIO0 through LPC1114_GPIO7 are allowed.</param>
        /// <param name="mode">LPC1114 GPIO pin mode.</param>
        public GPIO(ITransport spiagent, int pin, MODE mode)
        {
            // Validate parameters

            if (spiagent == null)
            {
                throw new NullReferenceException("SPI Agent Firmware transport handle is null");
            }

            if (!ValidPins.Contains(pin))
            {
                throw new ArgumentException("GPIO pin number is invalid");
            }

            if (mode >= MODE.SENTINEL)
            {
                throw new ArgumentException("GPIO mode parameter is invalid");
            }

            // Build the command message

            cmd.command = (int)Commands.SPIAGENT_CMD_CONFIGURE_GPIO;
            cmd.pin = pin;
            cmd.data = (int)mode;

            // Dispatch the command

            spiagent.Command(cmd, ref resp);

            // Handle errors

            if (resp.error != 0)
            {
                throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
            }

            mytransport = spiagent;
            mypin = pin;
        }

        /// <summary>
        /// Configure this GPIO pin's mode.
        /// </summary>
        /// <param name="mode">LPC1114 GPIO pin mode.</param>
        public void ConfigureMode(MODE mode)
        {
            // Build the command message

            cmd.command = (int)Commands.SPIAGENT_CMD_CONFIGURE_GPIO;
            cmd.pin = mypin;
            cmd.data = (int)mode;

            // Dispatch the command

            mytransport.Command(cmd, ref resp);

            // Handle errors

            if (resp.error != 0)
            {
                throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
            }
        }

        /// <summary>
        /// Configure this GPIO pin's interrupt mode.
        /// This only makes sense when running on the Raspberry Pi.
        /// </summary>
        /// <param name="intconfig">LPC1114 GPIO pin interrupt mode.</param>
        public void ConfigureInterrupt(INTERRUPT intconfig)
        {
            // Build the command message

            cmd.command = (int)Commands.SPIAGENT_CMD_CONFIGURE_GPIO_INTERRUPT;
            cmd.pin = mypin;
            cmd.data = (int)intconfig;

            // Dispatch the command

            mytransport.Command(cmd, ref resp);

            // Handle errors

            if (resp.error != 0)
            {
                throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
            }
        }

        /// <summary>
        /// This property gets or sets this LPC1114 GPIO pin's state.
        /// </summary>
        public bool state
        {
            get
            {
                // Build the command message

                cmd.command = (int)Commands.SPIAGENT_CMD_GET_GPIO;
                cmd.pin = mypin;
                cmd.data = 0;

                // Dispatch the command

                mytransport.Command(cmd, ref resp);

                // Handle errors

                if (resp.error != 0)
                {
                    throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
                }

                return (resp.data == 1);
            }

            set
            {
                // Build the command message

                cmd.command = (int)Commands.SPIAGENT_CMD_PUT_GPIO;
                cmd.pin = mypin;
                cmd.data = value ? 1 : 0;

                // Dispatch the command

                mytransport.Command(cmd, ref resp);

                // Handle errors

                if (resp.error != 0)
                {
                    throw new SPIAgent_Exception("SPI Agent Firmware returned error " + ((errno)resp.error).ToString());
                }
            }
        }
    }
}
